### ISSUE TITLE
Code review Feb 13

### ISSUE THREAD

**Issue by nataliemall** [nataliemall/crop_acreages_from_DPR_reports#1]
## Hello wonderful lab!

We'll skip the messy(er) part of processing the raw data, and instead download the handy compressed file of nice clean(ish) data (originally from the Pesticide Information Portal). 

 Step 1: Download the necessary files and put them in the same overall folder (or just clone the whole repository): 

```
calPUR_county_comparison.py
pur_and_county_data_retrieval.py
plotting_functions.py
acreage_calc_plots_single_irrigation_district.py

CA-crops-1980-2016.csv
county_commissioner_crop_types.csv
site_codes_with_crop_types.csv

```

Step 2: Download and unzip this file (sent to you via email since it's ~400mb). Put this unzipped folder into the into the same repo folder used in step 1.

```calPIP_PUR_crop_acreages.zip``` 
 
Step 3: Open the file below, choose a district, and run the script 
 
``` acreage_calc_plots_single_irrigation_district.py```

Bonus Step 4: Look at this script and run it for a little bit to see if it will start. Don't wait for the whole thing (it takes ~ 1 hour) . 

```
pur_data_with_comtrs.zip
compile_normalize_data_by_comtrs.py

```

 Here are my issues: 

1. The README file: Any ideas how I could simplify the readme? 
   Are the steps I want you to take doable? 

2. The script you could look at is acreage_calcs_plots_single_irrigation_district.py 

3. The main file I'm hoping to **eventually** make readable is compile_normalize_data_by_comtrs.py.  Right now it's a mess sorry. 

4. Commenting. I did not comment enough I am sorry in advance

5. Output files: I find myself caught in a situation where I'm afraid to reorganize because I have other files pointing to where things are right now.  Tips?

**Comment by jdherman** [comment]
Thanks Natalie. Making this clear for people is not easy, because the original dataset by itself has a lot of terminology and acronyms that you have to inherit.

1. The readme is at a good level of detail. The only thing I think might be helpful to add would be a short description of the format of the CSV data in each folder (on the flowchart). For example "Acres of tree crop for each comtrs" or something like that.

2. This script is readable enough. And it follows the format of changing some parameters at the top, and not changing anything else. A lot of the hard work is done in other functions though, like `retrieve_data_for_irrigation_district` ... which is maybe less interpretable.

The big picture is understandable. If anyone wanted to dig into the details it will be more tricky, but I think that's true of almost all data processing code (as opposed to a theoretical model). This could be a point worth discussing as a group.

I did run this, and it works just fine. I also like the `tqdm` thing to print the loop progress in the console.

3. Yea this script is tougher to understand. It is nice how the steps are broken up into separate functions, at least. And also good use of try-catch statements for error handling. To me it seems like most of the complexity is because of working with string data, and the need to alternate indexing by crop or by comtrs, etc. so I don't see an easy way around it.

5. Not sure what you mean. Is the issue with file paths breaking when things move? I didn't think there was anything wrong with the organization of output files (other than maybe the spaces in file and folder names).

I guess I'm not coming up with very specific suggestions here, because data processing scripts are almost always unavoidably complex compared to models (which can be more theoretical and designed from scratch). When you inherit a messy dataset, the code to process it will automatically be more complex. Group question: should we have different expectations for documentation and readability of data processing scripts? Do we just treat them as a black box, as long as we can understand the inputs and outputs? I'm not sure.

It wasn't an easy task to make sense of this data for your thesis, and you did a great job with it. And picked up some data science skills along the way.

**Comment by njmolla** [comment]
1.	 README is nicely laid out. The flowchart does a good job showing the steps and explaining the file structure. I’m not sure how much you can do to simplify it. 
2.	
 - The variable naming and commenting in this script are great. I think it’s good that you break things up into functions.
-	Some of the comments for the parameters are kind of weird because I’m assuming you commented them as such specifically for us (“keep this one equal to zero because I didn't give you the data”). I’m assuming you do sometimes mean for people to change parameters like `create_bar_chart` and `normalized`, so an explanation of what these are would be good in your final version
-	I like that you allow options to automatically look at all of the irrigation districts in a larger area
-	Running didn’t work for me ☹ It threw an error about not being able to find a file, maybe a Windows path issue?
3.	 
 - Agree that this file is definitely harder to understand. More comments would help, but I think that it also might just be too much to understand without a lot more extensive of a code review. 
-	Would be nice if the private functions (the ones used solely in other functions) are clearly delineated just so it’s clear what all the different levels of functions are
-	Not sure if I understand this but in the script part, you set `compilation` and `normalization`, but then have an if statement using those. Are they meant to be able to be changed?
-	You have a lot of commented out code. Not sure if that’s because those are things that you sometimes want or if it’s just old code. If it’s something you sometimes want, it would be better to have those as parameters you can set.
4.	Not sure if I understand or have any good suggestions for this. I guess if you just want to make it easier to change directory names, have them all as variables aggregated in the same part of the script so you can find them more easily? Or what I did is have functions pass paths to other functions so that you only have to define them in one function (not sure how applicable that is to you).

Overall I think you did a great job considering how complicated and messy the data processing you had to do was.

**Comment by mrlmautner** [comment]
1. README is clear and the order of operations is clear now that you have it matching the flowchart. You could be a little bit more detailed in some places, like insead of just "bad data", describe the type of cleaning you are doing.
2. In acreage_calcs_plots_single_irrigation_district.py, you can just import the whole scripts plotting_functions and the other one and you don't have to have multiple lines for each function. I wasn't able to run it because I don't have tqdm installed and it was taking forever to download.
Maybe instead of having the different lines commented out you could just include a text file somewhere that has the different options available and then just have one line with a comment that says "change this line" or something
3. Yeah, this one seems to be a little more complicated and needs more commenting overall.
4. Actually, the first script really didn't need much commenting because your variable names are really descriptive.
5. I wouldn't reorganize, especially since you are very clear in the flowchart about where everything goes.

**Comment by brobinson3** [comment]
1. I don’t think the ReadMe needs to be more simplified. If anything, you could expand on it by describing in more detail what each of the parts is/does. I like the flowchart!
2. I really like how simple the script is and that you rely on functions from other place to do a lot of the work for you. This makes your main script easy to understand and work on. I’m not sure why you need progress bars to print for each year (maybe instead just one progress bar showing the total progress?). I like the progress bar idea in general, but they flew by so fast that I didn’t know why they were there until the script ended and I read some of the text.
4. I liked the commenting on the main script. You made sure to say what sections were so that a user could understand how to operate the script. You really don't need much more there.
5. Figure out which organization makes the most sense, and then just bite the bullet, starting at your first script. Keep organizing until everything is in the correct new place and everything works. I don’t think there’s an easy way around this one (sorry). If you’re worried about breaking something and not being able to fix it, maybe work on a clone repository so you have the old one to go back to and look at for reference.
Overall, this whole project looks really complex and I'm impressed with how well you paired everything down and made it make sense.

**Comment by jscohen4** [comment]
1. The readme file is well put together. The flowchart looks complicated at first, but it’s actually pretty easy to follow. I think it would be near impossible to understand the flow of all the data without it. So it’s probably something useful for data-processing scripts. One think I think you should add to give a list of the packages that the repository needs (bumpy, pandas, tqdm, matplotlib, whatever else it might use ). I didn’t have tqdm installed initially, so the code gave me an error at first. 

2. First question after running the code is whether the figures get saved somewhere? I saw in some of your plotting scripts you used save_fig, but I wasn’t able to find where the figures were saving to. I liked that you listed out all the county options, but it seemed weird to have to comment in and out to do other counties. In general I feel like it might be better to just have that as an option a user could type in. It might get confusing if some of the process in running the code involves commenting and uncommenting things. I liked the 1 and 0 strategy for telling the script which options to do. Could also do True or False, but probably doesn’t matter. Also, could this file be called main.py (or is it not the main script?) P.S. the tqdm thing was dope. 
3. This file seems very specific to cleaning up the specific data files you’re working with, so it might be kinda hard to understand, even with a bunch of comments. One idea could be commenting what you’re doing to the data and especially why it needs to be done. 
4. In general in the  acreage_calcs_plots_single_irrigation_district.py I thought the commenting was good, concise, and not overwhelming.

**Comment by ekblad** [comment]
I agree with Jon Cohen's no. 2 and 3, and to elaborate slightly on his no. 3: an explanation of the normalization process, the actual calculations, in the readme would be helpful, since this is a choice that you were forced to make - metadata did not tell you why some sections reported more than 640 acres, correct?  i also think it warrants a longer comment than you have now in the code.  in general, i think that it is good practice to tell the story of where we get data (you do this well), what we are doing to it (you do this well), and if we are doing anything that degrades the information in any way and why we are doing that.  i am thinking about matching the crop codes from 1989 to 1990, which made me make choices about the data set that probably aren't exactly correct, but that I had to do to construct the time series.   another example of this might be when we downscale or lump or average, etc., simplifications that allow models to run better/faster since we don't need high resolution in all cases, but still a story that needs to be told since these simplifications alter the nature of the data we are working with.
